r/R/json.R

```diff
@@ -0,0 +1,139 @@
+  )
+)
+
+`arrow::json::ReadOptions` <- R6Class("arrow::json::ReadOptions", inherit = `arrow::Object`)
+`arrow::json::ParseOptions` <- R6Class("arrow::json::ParseOptions", inherit = `arrow::Object`)
+
+#' read options for the json reader
+#'
+#' @param use_threads Whether to use the global CPU thread pool
+#' @param block_size Block size we request from the IO layer; also determines the size of chunks when use_threads is `TRUE`. NB: if false, input must end with an empty line
+#'
+#' @export
+json_read_options <- function(use_threads = TRUE, block_size = 1048576L) {
+  shared_ptr(`arrow::json::ReadOptions`, json___ReadOptions__initialize(
+    list(
+      use_threads = use_threads,
+      block_size = block_size
+    )
+  ))
+}
+
+#' Parsing options
+#'
+#' @param newlines_in_values Whether objects may be printed across multiple lines (for example pretty printed),
+#'
+#' @export
+json_parse_options <- function(
+  newlines_in_values = FALSE
+){
+  shared_ptr(`arrow::json::ParseOptions`, json___ParseOptions__initialize(
+    list(
+      newlines_in_values = newlines_in_values
+    )
+  ))
+}
+
+#' Json table reader
+#'
+#' @param file file
+#' @param read_options, see [json_read_options()]
+#' @param parse_options, see [json_parse_options()]
+#' @param ... additional parameters.
+#'
+#' @export
+json_table_reader <- function(file,
+  read_options = json_read_options(),
+  parse_options = json_parse_options(),
+  ...
+){
+  UseMethod("json_table_reader")
+}
+
+#' @importFrom rlang abort
+#' @export
+json_table_reader.default <- function(file,
+  read_options = json_read_options(),
+  parse_options = json_parse_options(),
+  ...
+) {
+  abort("unsupported")
+}
+
+#' @export
+`json_table_reader.character` <- function(file,
+  read_options = json_read_options(),
+  parse_options = json_parse_options(),
+  ...
+){
+  json_table_reader(fs::path_abs(file),
+    read_options = read_options,
+    parse_options = parse_options,
+    ...
+  )
+}
+
+#' @export
+`json_table_reader.fs_path` <- function(file,
+  read_options = json_read_options(),
+  parse_options = json_parse_options(),
+  ...
+){
+  json_table_reader(ReadableFile(file),
+    read_options = read_options,
+    parse_options = parse_options,
+    ...
+  )
+}
+
+#' @export
+`json_table_reader.arrow::io::InputStream` <- function(file,
+  read_options = json_read_options(),
+  parse_options = json_parse_options(),
+  ...
+){
+  shared_ptr(`arrow::json::TableReader`,
+    json___TableReader__Make(file, read_options, parse_options)
+  )
+}
+
+#' @export
+`json_table_reader.arrow::json::TableReader` <- function(file,
+  read_options = json_read_options(),
+  parse_options = json_parse_options(),
+  ...
+){
+  file
+}
+
+#' Read json file into an arrow::Table
+#'
+#' Use arrow::json::TableReader from [json_table_reader()]
+#'
+#' @param ... Used to construct an arrow::json::TableReader
+#' @export
+read_json_arrow <- function(...) {
```

Comment:
For consistency, this should have an `as_tibble` argument right? (Happy to defer the `as_tibble` cleanup to a separate issue since it cuts across all the readers.)

Reply:
✅